feat(deleteReferences): Add dynamic parameters

michaelprichardson opened the pull request:
Added feature #1 

Can reference the masterId:
```
source: {
  collection: 'master',
},
targets: [{
    collection: 'detail1',
    foreignKey: 'masterId',
  },
  {
    collection: 'somecoll/{masterId}/detail2',
    foreignKey: 'masterId',
  },
],
```

Can reference the snapshot fields from the source collection (The keys of the field need to match the {} in the target otherwise it will skip it):
```
source: {
  collection: 'master',
},
targets: [{
    collection: 'detail1',
    foreignKey: 'masterId',
  },
  {
    collection: 'somecoll/{testId}/detail2',
    foreignKey: 'masterId',
  },
],
```

michaelprichardson committed:
feat(deleteReferences): Add dynamic parameters

michaelprichardson committed:
refactor(deleteReferences): Change target to use $

michaelprichardson committed:
test(deleteReferences): Add more test coverage

michaelprichardson committed:
test(deleteReferences): Fix test descriptions

michaelprichardson committed:
refactor(deleteReferences): Prettier formatting

nbransby commented on src/rules/deleteReferences.ts:
```diff
@@ -55,13 +55,33 @@ export function integrifyDeleteReferences(
+            context.params,
+            target.collection
+          );
+          target.collection = paramSwap.targetCollection;
```
this confused me for a minute would it be better to omit this...

nbransby commented on src/rules/deleteReferences.ts:
```diff
@@ -55,13 +55,33 @@ export function integrifyDeleteReferences(
+          // Replace the snapshot fields in the target collection
+          const fieldSwap = replaceReferenceWithFields(
+            snap.data(),
+            target.collection
```
...and use paramSwap.targetCollection here?

nbransby commented on src/rules/deleteReferences.ts:
```diff
@@ -55,13 +55,33 @@ export function integrifyDeleteReferences(
+          target.collection = fieldSwap.targetCollection;
+        } catch (error) {
+          throw new Error(error);
+        }
```
no point of catching an error and simply throwing it again

nbransby commented on test/functions/index.js:
```diff
@@ -62,6 +62,81 @@ module.exports.deleteReferencesToMaster = integrify({
+      foreignKey: 'masterId',
+    },
+    {
+      collection: 'somecoll/$masterId/detail2',
+      foreignKey: 'masterId',
```
in this example should the detail2 collection have a masterId field as its already in the path?

nbransby replied:
Seems strange to use a the variable $masterId which is not defined in the source path  - where is it coming from?

michaelprichardson replied:
The `masterId` is the id of the source collection. It would come from the path though. You think it would be easier to understand if we used `{...}` for path variables and `$...` for field variables?

nbransby replied:
if it comes from the source path where is it defined, shouldnt `$masterId` in the target throw an error because the source  path doesnt define the variable?

nbransby replied:
Is masterId a magic var created from 'Id' appended on collection name?

in the security rules syntax you would include the id in the source path if you wanted to reference it the rule:

`match /users/{uid} {` ...

Maybe we can do the same here?

```
source: {
    collection: 'master/{masterId}',
  },
  targets: [
    {
      collection: 'somecoll/$masterId/detail2',
       ...
    },
```

michaelprichardson replied:
```
const paramSwap = replaceReferenceWithFields(context.params, target.collection);
```
This line replaces the `$...` with the params, which masterId is part of. The next `replaceReferenceWithFields` replaces the `$..` with the fields. (Looking at it now the function name could be changed to some like `replaceReferenceWith(...)` to read better)

michaelprichardson replied:
Yeah, it has just been defined for us already and they chose `masterId` (line 34 in deleteReference.ts). We could use `collection: 'master/{masterId}'` to make it easier to understand too

nbransby replied:
so its a secret magic var caleld masterId! Really confusing as docs dont state this and their example collection is called master so you might conclude it comes from the collection name. 

Ok would make a lot of sense to require it to be defined in the source path and respects the name give to it

```
source: {
    collection: 'users/{userId}',
  },
  targets: [
    {
      collection: 'somecoll/$userId/detail2',
       ...
    },
```

michaelprichardson committed:
test(deleteReferences): Add tests for source inline primary key

michaelprichardson committed:
feat(deleteReferences): Specify primary key in source collection ref

michaelprichardson commented:
Changed the source collection to use:
```
source: {
  collection: 'users/{userId}',
},
targets: [{
    collection: 'somecoll/$userId/detail2',
    ...
  },
]
```

Added error checks for a missing primary key in the source collection. Added test for new functions and to test when primary key is left out from the source collection.

michaelprichardson committed:
feat(deleteReferences): Primary key is optional

michaelprichardson committed:
test(deleteReferences): Add test for optional primary key

michaelprichardson commented:
```
source: {
  collection: 'users/{userId}', <-- Optional now, will default to masterId as before
}
```
